apps/src/lib/node/ledger/shell/prepare_proposal.rs

```diff
@@ -77,6 +64,231 @@ mod prepare_block {
+                .to_bytes();
+            let tx_record = record::add(tx);
+
+            vec![tx_record]
+        }
+
+        /// Builds a batch of mempool transactions
+        fn build_mempool_txs(&mut self, txs: Vec<Vec<u8>>) -> Vec<TxRecord> {
+            // filter in half of the new txs from Tendermint, only keeping
+            // wrappers
+            let number_of_new_txs = 1 + txs.len() / 2;
+            txs.into_iter()
+                .take(number_of_new_txs)
+                .map(|tx_bytes| {
+                    if let Ok(Ok(TxType::Wrapper(_))) =
+                        Tx::try_from(tx_bytes.as_slice()).map(process_tx)
+                    {
+                        record::keep(tx_bytes)
+                    } else {
+                        record::remove(tx_bytes)
+                    }
+                })
+                .collect()
+        }
+
+        /// Builds a batch of DKG decrypted transactions
+        fn build_decrypted_txs(&mut self) -> Vec<TxRecord> {
+            // TODO: This should not be hardcoded
+            let privkey = <EllipticCurve as PairingEngine>::G2Affine::prime_subgroup_generator();
+
+            self.storage
+                .tx_queue
+                .iter()
+                .map(|tx| {
+                    Tx::from(match tx.decrypt(privkey) {
+                        Ok(tx) => DecryptedTx::Decrypted(tx),
+                        _ => DecryptedTx::Undecryptable(tx.clone()),
+                    })
+                    .to_bytes()
+                })
+                .map(record::add)
+                .collect()
+        }
+
+        /// Compresses a set of vote extensions into a single
+        /// [`VoteExtensionDigest`], whilst filtering invalid
+        /// `Signed<VoteExtension>` instances in the process
+        fn compress_vote_extensions(
+            &self,
+            vote_extensions: Vec<ExtendedVoteInfo>,
+        ) -> Option<VoteExtensionDigest> {
+            let events_epoch = self
+                .storage
+                .block
+                .pred_epochs
+                .get_epoch(self.storage.last_height)
+                .expect(
+                    "The epoch of the last block height should always be known",
+                );
+
+            let all_vote_extensions =
+                vote_extensions.into_iter().filter_map(|vote| {
+                    let vote_extension =
+                        Signed::<VoteExtension>::try_from_slice(
+                            &vote.vote_extension[..],
+                        )
+                        .map_err(|err| {
+                            tracing::error!(
+                                "Failed to deserialize signed vote extension: \
+                                 {}",
+                                err
+                            );
+                        })
+                        .ok()
+                        .and_then(|ext| {
+                            let ext_height = ext.data.block_height;
```

Comment:
This block can be simplified now that you have this https://github.com/anoma/namada/blob/5e6ea5ee57cdcc9acfc9e7a9a5119dd7f27a58f7/apps/src/lib/node/ledger/shell/vote_extensions.rs#L83